### PR TITLE
[Numpy] Add linalg.eig/eigh/eigvals/eigvalsh op

### PR DESCRIPTION
**Pull request by sjtuWangDing**
## Description ##
Add linalg.eig/eigh/eigvals/eigvalsh op

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: tests/python/unittest/test_numpy_op.py [hunk 1]
```diff
@@ -4291,6 +4291,377 @@ def newInvertibleMatrix_2D(shape, max_cond=4):
...
+            w_expected = _np.linalg.eigvalsh(a_np, UPLO)
+        except Exception as e:
+            print("a:", a_np)
+            print("a shape:", a_np.shape)
+            print(e)
+        else:
+            assert w.shape == w_expected.shape
+            assert_almost_equal(w, w_expected, rtol=rtol, atol=atol)
+
+    def newOrthonormalMatrix_2D(n):
```
Comment (reminisce) [inline tests/python/unittest/test_numpy_op.py]: Please add util functions to `mxnet.test_utils`. Also, please adopt python function's naming convention.

File: src/operator/numpy/linalg/np_eig-inl.h [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+        // Op forward implement.
+        EigOpForwardImpl<cpu>(a, w, v, req, &workspace, ctx.get_stream<cpu>());
+      } else {
+        std::vector<AType> a_vec(a.Size(), 0);
+        std::vector<WType> w_vec(w.Size(), 0);
+        std::vector<WType> v_vec(v.Size(), 0);
+        AType* a_cp_ptr = a_vec.data();
+        WType* w_cp_ptr = w_vec.data();
+        WType* v_cp_ptr = v_vec.data();
+        GpuCallbackCpuImpl(a, w, v, a_cp_ptr, w_cp_ptr, v_cp_ptr, &workspace, ctx, req);
```
Comment (haojin2) [inline src/operator/numpy/linalg/np_eig-inl.h]: suggested:
```c++
#if MXNET_USE_CUDA
template<typename AType, typename WType>
void GpuCallbackCpuImpl(...) {
}
#endif

// ...

if (xpu::kDevCPU) {
  // ...
} else {
#if MXNET_USE_CUDA
  // ...
  GpuCallbackCpuImpl(...);
#else
  LOG(FATAL) << "Please build with USE_CUDA=1 to enable GPU";
#endif
}
```

File: src/operator/numpy/linalg/np_eigvals.cc [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+      a_shape_vec[eig_ndim] = eig_shape[eig_ndim - 1];
+      mxnet::TShape a_shape(a_shape_vec.begin(), a_shape_vec.end());
+      SHAPE_ASSIGN_CHECK(*in_attrs, 0, a_shape);
+    }
+  }
+  return shape_is_known(*in_attrs) && shape_is_known(*out_attrs);
+}
+
+inline bool EigvalsOpType(const nnvm::NodeAttrs& attrs,
+                      std::vector<int>* in_attrs,
```
Comment (haojin2) [inline src/operator/numpy/linalg/np_eigvals.cc]: alignment

File: src/operator/numpy/linalg/np_eigvals-inl.h [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+  DMLC_DECLARE_PARAMETER(EigvalshParam) {
+    DMLC_DECLARE_FIELD(UPLO)
+    .set_default('L')
+    .describe("Specifies whether the calculation is done with the lower or upper triangular part.");
+  }
+};
+
+template<typename xpu>
+size_t EighEigvalshForwardWorkspaceSize(const TBlob& a,
+                                    const TBlob& w,
```
Comment (haojin2) [inline src/operator/numpy/linalg/np_eigvals-inl.h]: alignment

File: src/operator/numpy/linalg/np_eigvals.cc [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    // Calculate eig shape.
+    std::vector<int> eig_shape_vec(a_ndim - 1, -1);
+    for (int i = 0; i < a_ndim - 1; ++i) {
+      eig_shape_vec[i] = a_shape[i];
+    }
+    mxnet::TShape eig_shape(eig_shape_vec.begin(), eig_shape_vec.end());
+    SHAPE_ASSIGN_CHECK(*out_attrs, 0, eig_shape);
+  } else {
+    // Backward shape inference.
+    if (shape_is_known(eig_shape)) {
```
Comment (haojin2) [inline src/operator/numpy/linalg/np_eigvals.cc]: you could probably use
```c++
if (shape_is_known(a_shape)) {
  // code
} else if (shape_is_known(eig_shape)) {
  // code
}
```

File: src/operator/numpy/linalg/np_eig.cu [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+size_t EigForwardWorkspaceSize<gpu>(const TBlob& a,
+                                    const TBlob& w,
+                                    const TBlob& v,
+                                    const std::vector<OpReqType>& req) {
+  LOG(FATAL) << "EigForwardWorkspaceSize in gpu is unsupported";
+  return 0U;
+}
+
+template<>
+void EigOpForwardImpl<gpu>(const TBlob& a,
```
Comment (haojin2) [inline src/operator/numpy/linalg/np_eig.cu]: Those 2 functions are never called, you could probably delete them.

File: python/mxnet/symbol/numpy/linalg.py [hunk 7]
```diff
@@ -496,3 +497,265 @@ def tensorsolve(a, b, axes=None):
...
+    if `Q` is orthogonal, then ``Q * A * Q.T`` has the same eigenvalues as
+    ``A``:
+    >>> from numpy import linalg as LA
+    >>> x = np.random.random()
+    >>> Q = np.array([[np.cos(x), -np.sin(x)], [np.sin(x), np.cos(x)]])
+    >>> LA.norm(Q[0, :]), LA.norm(Q[1, :]), np.dot(Q[0, :],Q[1, :])
+    (1.0, 1.0, 0.0)
+
+    Now multiply a diagonal matrix by ``Q`` on one side and by ``Q.T`` on the other:
+    >>> D = np.diag((-1,1))
```
Comment (haojin2) [inline python/mxnet/symbol/numpy/linalg.py]: No need for examples for symbolic APIs.

File: src/operator/numpy/linalg/np_eigvals.cc [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+  return shape_is_known(*in_attrs) && shape_is_known(*out_attrs);
+}
+
+inline bool EigvalsOpType(const nnvm::NodeAttrs& attrs,
+                      std::vector<int>* in_attrs,
+                      std::vector<int>* out_attrs) {
+  CHECK_EQ(in_attrs->size(), 1U);
+  CHECK_EQ(out_attrs->size(), 1U);
+  int a_type = in_attrs->at(0);
+  // unsupport float16
```
Comment (haojin2) [inline src/operator/numpy/linalg/np_eigvals.cc]: Does the 2 ops support integer inputs?
If not, then I would suggest the following code structure:
```c++
if (common::is_float(a_type)) {
  // raise error for fp16
  TYPE_ASSIGN_CHECK(*out_attrs, 0, in_attrs->at(0));
} else {
  // raise error
}
```

File: src/operator/numpy/linalg/np_eig.cc [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+      SHAPE_ASSIGN_CHECK(*out_attrs, 0, eig_shape);
+    }
+  }
+  return shape_is_known(*in_attrs) && shape_is_known(*out_attrs);
+}
+
+inline bool EigOpType(const nnvm::NodeAttrs& attrs,
+                      std::vector<int>* in_attrs,
+                      std::vector<int>* out_attrs) {
+  CHECK_EQ(in_attrs->size(), 1U);
```
Comment (haojin2) [inline src/operator/numpy/linalg/np_eig.cc]: see the comment below on the other InferType function.

File: src/operator/numpy/linalg/np_eigvals-inl.h [hunk 10]
```diff
@@ -0,0 +1,432 @@
...
+                               wr.dptr_, wi.dptr_, \
+                               vl.dptr_, ldvl, \
+                               vr.dptr_, ldvr, \
+                               work_array.dptr_, lwork)); \
+  CHECK_LE(res, 0) << #fname << "the QR algorithm failed to compute all the" \
+    << "eigenvalues, and no eigenvectors have been computed; elements " \
+    << res + 1 << ":N" << " of WR and WI contain eigenvalues which have converged"; \
+  CHECK_GE(res, 0) << #fname << ": the " << -res \
+    << "-th argument had an illegal value"; \
+}
```
Comment (haojin2) [inline src/operator/numpy/linalg/np_eigvals-inl.h]: blank line below.

File: src/operator/numpy/linalg/np_eigvals.cc (hunk 8 above)
Comment (sjtuWangDing) [reply src/operator/numpy/linalg/np_eigvals.cc]: These 2 ops support integer inputs in numpy.

**Pull request merged**